scipy/stats/_stats_py.py

```diff
@@ -293,9 +307,13 @@ def hmean(a, axis=0, dtype=None, *, weights=None):
     if weights is not None:
         weights = np.asarray(weights, dtype=dtype)
 
-    if not np.all(a >= 0):
-        message = ("The harmonic mean is only defined if all elements are greater "
-                   "than or equal to zero; otherwise, the result is NaN.")
+    negative_mask = a < 0
+    if np.any(negative_mask):
+        # `where` avoids having to be careful about dtypes and will work with
+        # JAX. This is the exceptional case, so it's OK to be a little slower.
+        a = np.where(negative_mask, np.nan, a)
```

Comment:
This (arguably) fixes a mistake in the last PR - we emitted a warning with negative observations, but returned the results according to the usual formula. That works for `gmean` because the log of a negative number is NaN, but it doesn't necessarily work for these. 

That said, I think one could define `pmean` (at least) more generally. As-is, `pmean` with `p=1` (i.e. THE mean) is undefined for negative input. Presumably this was discussed in the original PR, but it's worth mentioning again.